File: packages/navigation/ios/ExpoGaodeMapNaviViewModule.swift
//
//  ExpoGaodeMapNaviViewModule.swift
//  expo-gaode-map-navigation
//
//  
//

import ExpoModulesCore
import UIKit

public class ExpoGaodeMapNaviViewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoGaodeMapNaviView")
    
    // 定义视图
    View(ExpoGaodeMapNaviView.self) {
      // 事件定义
      Events(
        "onNavigationReady",
        "onNavigationStarted",
        "onNavigationFailed",
        "onNavigationEnded",
        "onLocationUpdate",
        "onNavigationText",
        "onArriveDestination",
        "onRouteCalculated",
        "onRouteRecalculate",
        "onWayPointArrived",
        "onGpsStatusChanged",
        "onNavigationInfoUpdate",
        "onGpsSignalWeak",
        "onNavigationVisualStateUpdate",
        "onLaneInfoUpdate",
        "onTrafficStatusesUpdate"
      )
      
      // 属性定义
      Prop("naviType") { (view: ExpoGaodeMapNaviView, value: Int) in
        view.naviType = value
      }
      
      Prop("enableVoice") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.enableVoice = value
      }
      
      Prop("showCamera") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showCamera = value
      }

      Prop("carImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.carImageSource = value
      }

      Prop("carImageSize") { (view: ExpoGaodeMapNaviView, value: [String: Double]?) in
        let width = CGFloat(value?["width"] ?? 0)
        let height = CGFloat(value?["height"] ?? 0)
        if width > 0, height > 0 {
          view.carImageSize = CGSize(width: width, height: height)
        } else {
          view.carImageSize = nil
        }
      }

      Prop("carCompassImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.carCompassImageSource = value
      }

      Prop("startPointImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.startPointImageSource = value
      }

      Prop("wayPointImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.wayPointImageSource = value
      }

      Prop("customWaypointMarkers") { (view: ExpoGaodeMapNaviView, value: [[String: Any]]?) in
        view.customWaypointMarkerPayloads = value
      }

      Prop("endPointImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.endPointImageSource = value
      }

      Prop("cameraImage") { (view: ExpoGaodeMapNaviView, value: String?) in
        view.cameraImageSource = value
      }
      
      Prop("autoLockCar") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.autoLockCar = value
      }
      
      Prop("autoChangeZoom") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.autoChangeZoom = value
      }
      
      Prop("trafficLayerEnabled") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.trafficLayerEnabled = value
      }
      
      Prop("realCrossDisplay") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.realCrossDisplay = value
      }
      
      Prop("naviMode") { (view: ExpoGaodeMapNaviView, value: Int) in
        view.naviMode = value
      }
      
      Prop("showMode") { (view: ExpoGaodeMapNaviView, value: Int) in
        view.showMode = value
      }
      
      Prop("isNightMode") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.isNightMode = value
      }
      
      // iOS 特有属性
      Prop("showRoute") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showRoute = value
      }
      
      Prop("naviArrowVisible") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showTurnArrow = value
      }
      
      Prop("showTrafficBar") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showTrafficBar = value
      }

      Prop("trafficBarFrame") { (view: ExpoGaodeMapNaviView, value: [String: Double]?) in
        let x = CGFloat(value?["x"] ?? 0)
        let y = CGFloat(value?["y"] ?? 0)
        let width = CGFloat(value?["width"] ?? 0)
        let height = CGFloat(value?["height"] ?? 0)
        view.trafficBarFrame = CGRect(x: x, y: y, width: width, height: height)
      }

      Prop("trafficBarColors") { (view: ExpoGaodeMapNaviView, value: [String: Any]?) in
        view.trafficBarColors = value
      }
      
      Prop("showBrowseRouteButton") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showBrowseRouteButton = value
      }
      
      Prop("showMoreButton") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showMoreButton = value
      }
      
      Prop("showTrafficButton") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showTrafficButton = value
      }

      Prop("showBackupRoute") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showBackupRoute = value
      }

      Prop("showEagleMap") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showEagleMap = value
      }
      
      Prop("showUIElements") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.applyShowUIElements(value)
      }
      
      Prop("showGreyAfterPass") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showGreyAfterPass = value
      }
      
      Prop("showVectorline") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showVectorline = value
      }
      
      Prop("showTrafficLights") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showTrafficLights = value
      }

      Prop("showCompassEnabled") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showCompassEnabled = value
      }

      Prop("showDriveCongestion") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showDriveCongestion = value
      }

      Prop("showTrafficLightView") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.showTrafficLightView = value
      }

      Prop("hideNativeTopInfoLayout") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.hideNativeTopInfoLayout = value
      }

      Prop("hideNativeLaneInfoLayout") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.hideNativeLaneInfoLayout = value
      }

      // Android-only prop, keep a no-op mapping on iOS for cross-platform prop compatibility.
      Prop("androidBackgroundNavigationNotificationEnabled") { (_: ExpoGaodeMapNaviView, _: Bool) in
      }

      Prop("iosLiveActivityEnabled") { (view: ExpoGaodeMapNaviView, value: Bool) in
        view.iosLiveActivityEnabled = value
      }
      
      Prop("mapViewModeType") { (view: ExpoGaodeMapNaviView, value: Int) in
        view.mapViewModeType = value
      }
      
      Prop("lineWidth") { (view: ExpoGaodeMapNaviView, value: Double) in
        view.lineWidth = CGFloat(value)
      }

      Prop("driveViewEdgePadding") { (view: ExpoGaodeMapNaviView, value: [String: Double]?) in
        let top = CGFloat(value?["top"] ?? 0)
        let left = CGFloat(value?["left"] ?? 0)
        let bottom = CGFloat(value?["bottom"] ?? 0)
        let right = CGFloat(value?["right"] ?? 0)
        view.driveViewEdgePadding = UIEdgeInsets(top: top, left: left, bottom: bottom, right: right)
      }

      Prop("screenAnchor") { (view: ExpoGaodeMapNaviView, value: [String: Double]?) in
        let x = CGFloat(value?["x"] ?? 0)
        let y = CGFloat(value?["y"] ?? 0)
        view.screenAnchor = CGPoint(x: x, y: y)
      }
      
      // 方法定义
      AsyncFunction("startNavigation") { (view: ExpoGaodeMapNaviView, startLat: Double, startLng: Double, endLat: Double, endLng: Double, promise: Promise) in
        view.startNavigation(startLat: startLat, startLng: startLng, endLat: endLat, endLng: endLng, promise: promise)
      }

      AsyncFunction("startNavigationWithIndependentPath") { (view: ExpoGaodeMapNaviView, token: Int, routeId: Int?, routeIndex: Int?, naviType: Int?, promise: Promise) in
        view.startNavigationWithIndependentPath(token: token, routeId: routeId, routeIndex: routeIndex, naviType: naviType, promise: promise)
      }
      
      AsyncFunction("stopNavigation") { (view: ExpoGaodeMapNaviView, promise: Promise) in
        view.stopNavigation(promise: promise)
      }
      
      AsyncFunction("playCustomTTS") { (view: ExpoGaodeMapNaviView, text: String, forcePlay: Bool, promise: Promise) in
        view.playCustomTTS(text: text, forcePlay: forcePlay, promise: promise)
      }
    }
  }
}

File: packages/navigation/android/src/main/java/expo/modules/gaodemap/navigation/ExpoGaodeMapNaviViewModule.kt
package expo.modules.gaodemap.navigation

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * 高德导航视图 Module
 */
class ExpoGaodeMapNaviViewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoGaodeMapNaviView")

    OnActivityEntersForeground {
      ExpoGaodeMapNaviView.resumeActiveViews()
    }

    OnActivityEntersBackground {
      ExpoGaodeMapNaviView.pauseActiveViews()
    }

    OnActivityDestroys {
      ExpoGaodeMapNaviView.destroyActiveViews()
    }

    View(ExpoGaodeMapNaviView::class) {
      Events(
        "onNavigationReady",
        "onNavigationStarted",
        "onNavigationFailed",
        "onNavigationEnded",
        "onLocationUpdate",
        "onNavigationText",
        "onArriveDestination",
        "onRouteCalculated",
        "onRouteRecalculate",
        "onWayPointArrived",
        "onGpsStatusChanged",
        "onNavigationInfoUpdate",
        "onGpsSignalWeak",
        "onNavigationVisualStateUpdate",
        "onLaneInfoUpdate",
        "onTrafficStatusesUpdate"
      )

      OnViewDestroys { view: ExpoGaodeMapNaviView ->
        view.onDestroy()
      }

      // Props
      Prop<Int>("naviType") { view, type ->
        view.applyNaviType(type)
      }
      
      Prop<Boolean>("enableVoice") { view, enabled ->
        view.applyEnableVoice(enabled)
      }

      Prop<String?>("carImage") { view, uri ->
        view.applyCarImage(uri)
      }

      Prop<Map<String, Double>?>("carImageSize") { view, value ->
        val width = value?.get("width")
        val height = value?.get("height")
        view.applyCarImageSize(width, height)
      }

      Prop<String?>("fourCornersImage") { view, uri ->
        view.applyFourCornersImage(uri)
      }

      Prop<String?>("startPointImage") { view, uri ->
        view.applyStartPointImage(uri)
      }

      Prop<String?>("wayPointImage") { view, uri ->
        view.applyWayPointImage(uri)
      }

      Prop<List<Map<String, Any?>>?>("customWaypointMarkers") { view, markers ->
        view.applyCustomWaypointMarkers(markers)
      }

      Prop<String?>("endPointImage") { view, uri ->
        view.applyEndPointImage(uri)
      }
      
      Prop<Boolean>("autoLockCar") { view, enabled ->
        view.applyAutoLockCar(enabled)
      }
      
      Prop<Boolean>("autoChangeZoom") { view, enabled ->
        view.applyAutoChangeZoom(enabled)
      }
      
      Prop<Boolean>("trafficLayerEnabled") { view, enabled ->
        view.applyTrafficLayerEnabled(enabled)
      }
      
      Prop<Boolean>("realCrossDisplay") { view, enabled ->
        view.applyRealCrossDisplay(enabled)
      }
      
      Prop<Int>("naviMode") { view, mode ->
        view.applyNaviMode(mode)
      }
      
      Prop<Int>("showMode") { view, mode ->
        view.applyShowMode(mode)
      }
      
      Prop<Boolean>("isNightMode") { view, enabled ->
        view.applyNightMode(enabled)
      }

      Prop<Int>("mapViewModeType") { view, mode ->
        view.applyMapViewModeType(mode)
      }
      
      Prop<Boolean>("carOverlayVisible") { view, visible ->
        view.applyCarOverlayVisible(visible)
      }
      
      Prop<Boolean>("showTrafficLights") { view, visible ->
        view.applyTrafficLightsVisible(visible)
      }

      Prop<Boolean>("showCompassEnabled") {view, enabled ->
          view.applyShowCompassEnabled(enabled)
      }

      Prop<Boolean>("naviStatusBarEnabled") { view, enabled ->
        view.applyNaviStatusBarEnabled(enabled)
      }
      
      Prop<Map<String, Boolean>?>("routeMarkerVisible") { view, config ->
        config?.let {
          val showStartEndVia = it["showStartEndVia"] as? Boolean ?: true
          val showFootFerry = it["showFootFerry"] as? Boolean ?: true
          val showForbidden = it["showForbidden"] as? Boolean ?: true
          val showRouteStartIcon = it["showRouteStartIcon"] as? Boolean ?: true
          val showRouteEndIcon = it["showRouteEndIcon"] as? Boolean ?: true
          view.applyRouteMarkerVisible(showStartEndVia, showFootFerry, showForbidden, showRouteStartIcon, showRouteEndIcon)
        }
      }
      
      // 路线转向箭头可见性控制
      Prop<Boolean>("naviArrowVisible") { view, visible ->
        view.applyNaviArrowVisible(visible)
      }

      Prop<Boolean>("laneInfoVisible") { view, visible ->
        view.applyLaneInfoVisible(visible)
      }

      Prop<Boolean>("modeCrossDisplay") { view, visible ->
        view.applyModeCrossDisplay(visible)
      }

      Prop<Boolean>("eyrieCrossDisplay") { view, visible ->
        view.applyEyrieCrossDisplay(visible)
      }

      Prop<Boolean>("secondActionVisible") { view, visible ->
        view.applySecondActionVisible(visible)
      }

      Prop<Boolean>("backupOverlayVisible") { view, visible ->
        view.applyBackupOverlayVisible(visible)
      }
      
      // 是否显示拥堵气泡
      Prop<Boolean>("showDriveCongestion") { view, show ->
        view.applyShowDriveCongestion(show)
      }
      
      // 是否显示红绿灯倒计时气泡
      Prop<Boolean>("showTrafficLightView") { view, show ->
        view.applyShowTrafficLightView(show)
      }

     //设置是否为骑步行视图
        Prop<Boolean>("isNaviTravelView") { view, isRideStepView ->
            view.applyIsNaviTravelView(isRideStepView)
        }

      Prop<Double?>("androidStatusBarPaddingTop") { view, topDp ->
        view.applyAndroidStatusBarPaddingTop(topDp)
      }

      Prop<Int>("lockZoom") { view, level ->
        view.applyLockZoom(level)
      }

      Prop<Int>("lockTilt") { view, tilt ->
        view.applyLockTilt(tilt)
      }

      Prop<Boolean>("eagleMapVisible") { view, visible ->
        view.applyEagleMapVisible(visible)
      }

      Prop<Map<String, Double>?>("pointToCenter") { view, value ->
        val x = value?.get("x") ?: 0.0
        val y = value?.get("y") ?: 0.0
        view.applyPointToCenter(x, y)
      }

      Prop<Boolean>("hideNativeTopInfoLayout") { view, hidden ->
        view.applyHideNativeTopInfoLayout(hidden)
      }

      Prop<Boolean>("showCamera") { view, enabled ->
          view.applyShowCamera(enabled)
      }

      Prop<Boolean>("showUIElements") { view, visible ->
        view.applyShowUIElements(visible)
      }

      Prop<Boolean>("androidBackgroundNavigationNotificationEnabled") { view, enabled ->
        view.applyAndroidBackgroundNavigationNotificationEnabled(enabled)
      }

      // iOS-only prop, keep a no-op mapping on Android for cross-platform prop compatibility.
      Prop<Boolean>("iosLiveActivityEnabled") { _: ExpoGaodeMapNaviView, _: Boolean ->
      }

      Prop<Boolean>("showTrafficBar") { view, enabled ->
        view.applyAndroidTrafficBarEnabled(enabled)
      }

      Prop<Boolean>("showTrafficButton"){ view, enabled ->
          view.applyShowTrafficButton(enabled)
      }

      Prop<Boolean>("showBrowseRouteButton") { view, enabled ->
          view.applyShowBrowseRouteButton(enabled)
      }

        Prop<Boolean>("showVectorline"){ view, enabled ->
            view.applyShowVectorline(enabled)
        }

        Prop<Boolean>("showGreyAfterPass") { view, enabled ->
            view.applyShowGreyAfterPass(enabled)
        }

      // 方法
      AsyncFunction("startNavigation") { view: ExpoGaodeMapNaviView, startLat: Double, startLng: Double, endLat: Double, endLng: Double, promise: expo.modules.kotlin.Promise ->
        view.startNavigation(startLat, startLng, endLat, endLng, promise)
      }

      AsyncFunction("startNavigationWithIndependentPath") { view: ExpoGaodeMapNaviView, token: Int, routeId: Int?, routeIndex: Int?, naviType: Int?, promise: expo.modules.kotlin.Promise ->
        view.startNavigationWithIndependentPath(token, routeId, routeIndex, naviType, promise)
      }
      
      AsyncFunction("stopNavigation") { view: ExpoGaodeMapNaviView, promise: expo.modules.kotlin.Promise ->
        view.stopNavigation(promise)
      }

      AsyncFunction("playCustomTTS") { view: ExpoGaodeMapNaviView, text: String, forcePlay: Boolean, promise: expo.modules.kotlin.Promise ->
        val success = view.playCustomTTS(text, forcePlay)
        if (success) {
          promise.resolve(
            mapOf(
              "success" to true
            )
          )
        } else {
          promise.reject("PLAY_TTS_FAILED", "当前场景暂不支持或正在播报其他导航语音", null)
        }
      }
    }
  }
}
